pkg/runtime/depot.go

```diff
@@ -270,16 +273,44 @@ func (d *depot) Undeploy(id strfmt.UUID, relativeSrc, path string) error {
 	if !ok {
 		return errs.New("deployment for %s not found in depot", id)
 	}
-	deploy := sliceutils.Filter(deployments, func(d deployment) bool { return d.Path == path })
-	if len(deploy) != 1 {
+	deployments = sliceutils.Filter(deployments, func(d deployment) bool { return d.Path == path })
+	if len(deployments) != 1 {
 		return errs.New("no deployment found for %s in depot", path)
 	}
 
+	// Determine if there are any files provided by another artifact that will need to be re-linked
+	// or re-copied after this artifact is uninstalled.
+	relinks, err := d.getSharedFilesToReLink(id, relativeSrc, path)
+	if err != nil {
+		return errs.Wrap(err, "failed to get shared files")
+	}
+	sharedFiles := make([]string, 0)
+	for file := range relinks {
+		sharedFiles = append(sharedFiles, file)
+	}
+
 	// Perform uninstall based on deployment type
-	if err := smartlink.UnlinkContents(filepath.Join(d.Path(id), relativeSrc), path); err != nil {
+	if err := smartlink.UnlinkContents(filepath.Join(d.Path(id), relativeSrc), path, sharedFiles...); err != nil {
```

Comment:
It doesn't seem worth "ignoring" the shared files here. You're still removing it on line 299. Can't we drop the ignore mechanic and line 299 through 301 and still have the same result?

Reply:
`UnlinkContents()` will remove paths and also empty directories, so we need to notify it of paths to ignore. Then, we need to remove an existing link before re-linking because `Link()` won't re-link if the target already exists.

That is the reason for why I did what I did.

Reply:
Yeah I think having it create directories feels more appropriate. Not deleting the directories is how git operates, so I think we can do the same.